### PR TITLE
chore: fix all unit tests  with multiplexed sessions.

### PR DESCRIPTION
**Pull request by arpan14**
Tests have been parameterized or skipped when test runners pass the env variable `GOOGLE_CLOUD_SPANNER_ENABLE_MULTIPLEXED_SESSIONS`. We presently have created separate test runners (unit, emulator and integration tests) which run during pre-submits and test the multiplexed session code path.

**Pending Stuff**

1. `ChannelUsageTest` has been disabled (with a followup TODO) when multiplexed sessions is enabled. We are coordinating with GRPC team to see how this can be enabled. This is the only test which currently exudes a flaky behavior very frequently.

### CODE REVIEW

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPool.java
Comment (arpan14) [inline google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPool.java:1796]: @olavloite this is something that I have added for the time being. But I am still debugging some more race conditions which is causing flaky failures in `SpannerGaxRetryTest`. I think there is still a scope to improve synchronization of the `SettableApiFuture` without causing deadlocks.

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPoolOptions.java
Comment (arpan14) [inline google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPoolOptions.java:287]: This is made public since its needed for connection API tests.

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPool.java
Comment (arpan14) [inline google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPool.java:1182]: We don't really need to save `SettableApiFuture<SessionReference>` per `MultiplexedSessionFutureWrapper`. This can directly be obtained from the global reference.

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPool.java
Comment (arpan14) [inline google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPool.java:3685]: @olavloite I think the order here does not matter. i.e it is safe to first do `poolMaintainer.close();` and then propagate the exceptions to all the waiting threads.

File: google-cloud-spanner/src/test/java/com/google/cloud/spanner/ChannelUsageTest.java
Comment (olavloite) [inline google-cloud-spanner/src/test/java/com/google/cloud/spanner/ChannelUsageTest.java:226]: This test works for the normal channel pool, it is only with Gcp pool that it fails.

```suggestion
          enableGcpPool && isMultiplexedSessionsEnabled(spanner));
```

File: google-cloud-spanner/src/test/java/com/google/cloud/spanner/ChannelUsageTest.java
Comment (olavloite) [inline google-cloud-spanner/src/test/java/com/google/cloud/spanner/ChannelUsageTest.java:240]:
```suggestion
          enableGcpPool && isMultiplexedSessionsEnabled(spanner));
```

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionClient.java
Comment (olavloite) [inline google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionClient.java:270]: nit:
```suggestion
   * SessionConsumer#onSessionCreateFailure(Throwable, int)} call with the error.
```

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPool.java
Comment (olavloite) [reply google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPool.java:1789]: Is it possible that multiple threads will call `MultiplexedSessionFuture#get()` (so this method) before it has been initialized? If so, then this would need its own lock.

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPool.java
Comment (olavloite) [inline google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPool.java:1793]: nit: I think that you only need to take the lock for the `incrementNumSessionsInuse(true);` call. The rest does not need the session pool lock.

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPool.java
Comment (olavloite) [inline google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPool.java:3681]: nit: use `asSpannerException(t)` instead of `newSpannerException(t)`. That will keep the original exception if it was already a `SpannerException`.

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPool.java
Comment (olavloite) [reply google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPool.java:3685]: Yes, I think so too.

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPoolOptions.java
Comment (olavloite) [reply google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPoolOptions.java:287]: nit: can we mark it with `@InternalApi` until the setter has also been made public? Or maybe `@VisibleForTesting`?

File: google-cloud-spanner/src/test/java/com/google/cloud/spanner/DatabaseClientImplTest.java
Comment (olavloite) [inline google-cloud-spanner/src/test/java/com/google/cloud/spanner/DatabaseClientImplTest.java:3016]: I don't think you need the `if` statement. `createSessions` is never used by the regular session pool, so you can just always add the exception for that as well.

File: google-cloud-spanner/src/test/java/com/google/cloud/spanner/DatabaseClientImplTest.java
Comment (olavloite) [inline google-cloud-spanner/src/test/java/com/google/cloud/spanner/DatabaseClientImplTest.java:3686]: I don't think you need separate tests for these conditions. `BatchCreateSessions` and `CreateSession` are both exclusively used by respectively non-multiplexed-sessions and multiplexed-sessions. That means that you can just register the same error for both, and depending on whether multiplexed sessions are enabled or not, one of the RPCs will be called, and the behavior should be the same in both cases.

File: google-cloud-spanner/src/test/java/com/google/cloud/spanner/IntegrationTestWithClosedSessionsEnv.java
Comment (olavloite) [inline google-cloud-spanner/src/test/java/com/google/cloud/spanner/IntegrationTestWithClosedSessionsEnv.java:91]: I don't think this is ever used, as the integration test that uses this test env is (correctly) skipped for multiplexed sessions. And if it is used, what does this then mean?

File: google-cloud-spanner/src/test/java/com/google/cloud/spanner/MockSpannerServiceImpl.java
Comment (olavloite) [inline google-cloud-spanner/src/test/java/com/google/cloud/spanner/MockSpannerServiceImpl.java:588]: This does not appear to be used. The only place where we need to check whether a session is multiplexed, is when a transaction is being executed, and that point the code just checks whether the `isMultiplexed` flag is set on that session. So I think we can safely store the multiplexed sessions in the same map as the normal sessions. That also makes methods like `listSessions` and `deleteSession` work (I think they will exclude any multiplexed session in this implementation).

File: google-cloud-spanner/src/test/java/com/google/cloud/spanner/ReadAsyncTest.java
Comment (olavloite) [inline google-cloud-spanner/src/test/java/com/google/cloud/spanner/ReadAsyncTest.java:188]: Same here: I don't think you need the if statement.

File: google-cloud-spanner/src/test/java/com/google/cloud/spanner/RetryOnInvalidatedSessionTest.java
Comment (olavloite) [inline google-cloud-spanner/src/test/java/com/google/cloud/spanner/RetryOnInvalidatedSessionTest.java:233]: Note: I also have a large update (cleanup) of this test pending in https://github.com/googleapis/java-spanner/pull/3040.

Once this PR has been merged, I'll rebase my PR and update it so it executes this test for the cases that still use regular sessions, even when multiplexed sessions are enabled (i.e. for read/write transactions). That will ensure that the session retry continues to work for those cases as well after enabling multiplexed sessions.

File: google-cloud-spanner/src/test/java/com/google/cloud/spanner/SessionPoolTest.java
Comment (olavloite) [inline google-cloud-spanner/src/test/java/com/google/cloud/spanner/SessionPoolTest.java:267]: here also: I think that you can remove the if statement, as it does not interfere with the non-multiplexed case

File: google-cloud-spanner/src/test/java/com/google/cloud/spanner/SessionPoolTest.java
Comment (olavloite) [inline google-cloud-spanner/src/test/java/com/google/cloud/spanner/SessionPoolTest.java:2189]: remove if?

File: google-cloud-spanner/src/test/java/com/google/cloud/spanner/SpanTest.java
Comment (olavloite) [inline google-cloud-spanner/src/test/java/com/google/cloud/spanner/SpanTest.java:204]: Is this needed in order to prevent it from being flaky? (It probably doesn't hurt, but it seems like a surprising change)

File: google-cloud-spanner/src/test/java/com/google/cloud/spanner/SpannerGaxRetryTest.java
Comment (olavloite) [inline google-cloud-spanner/src/test/java/com/google/cloud/spanner/SpannerGaxRetryTest.java:218]: In order to replicate the original behavior of this test, you should set `mockSpanner.setCreateSessionExecutionTime(ONE_SECOND)`.

File: google-cloud-spanner/src/test/java/com/google/cloud/spanner/SpannerGaxRetryTest.java
Comment (olavloite) [inline google-cloud-spanner/src/test/java/com/google/cloud/spanner/SpannerGaxRetryTest.java:138]: This test verifies the behavior of the client when a query is blocked while waiting for the session pool to have a session available. This changes that. Is there something in this test that does not work with multiplexed sessions if this wait time is not specified?

**Review by olavloite — APPROVED** [review]

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPoolOptions.java
Comment (arpan14) [reply google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPoolOptions.java:287]: Done

File: google-cloud-spanner/src/test/java/com/google/cloud/spanner/ChannelUsageTest.java
Comment (arpan14) [reply google-cloud-spanner/src/test/java/com/google/cloud/spanner/ChannelUsageTest.java:226]: I had originally done this. But it still introduced a bit of flaky behavior. I will be following up to re-enabled this test pending some investigation.

File: google-cloud-spanner/src/test/java/com/google/cloud/spanner/ChannelUsageTest.java
Comment (arpan14) [reply google-cloud-spanner/src/test/java/com/google/cloud/spanner/ChannelUsageTest.java:240]: I had originally done this. But it still introduced a bit of flaky behavior even without gcp-pool. I will be following up to re-enable this test pending some investigation.

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPool.java
Comment (arpan14) [reply google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPool.java:1793]: Yeah, I removed the lock from here since we are acquiring the lock again within the `incrementNumSessionsInUse` method.

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPool.java
Comment (arpan14) [reply google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPool.java:3681]: Done

File: google-cloud-spanner/src/test/java/com/google/cloud/spanner/DatabaseClientImplTest.java
Comment (arpan14) [reply google-cloud-spanner/src/test/java/com/google/cloud/spanner/DatabaseClientImplTest.java:3686]: Right, modified to re-use the same test.

File: google-cloud-spanner/src/test/java/com/google/cloud/spanner/IntegrationTestWithClosedSessionsEnv.java
Comment (arpan14) [reply google-cloud-spanner/src/test/java/com/google/cloud/spanner/IntegrationTestWithClosedSessionsEnv.java:91]: No, this is still used in the regular session call path. We have replaced `getSession()` method with `getMultiplexedSession ` in `DatabaseClientImpl` class for all RO transactions in this PR. So if we don't add this override behaviour, those integration tests will fail.

File: google-cloud-spanner/src/test/java/com/google/cloud/spanner/MockSpannerServiceImpl.java
Comment (arpan14) [reply google-cloud-spanner/src/test/java/com/google/cloud/spanner/MockSpannerServiceImpl.java:588]: Actually a couple of tests assert the size of `sessions` list. So if we add the multiplexed session in the same list, it will increase the size by 1 and then we would have to modify all those tests to work with a different list size (if-else) logic. I felt changing it here is a cleaner solution than modifying different tests.

* google-cloud-spanner/src/test/java/com/google/cloud/spanner/connection/ConnectionTest.java
* google-cloud-spanner/src/test/java/com/google/cloud/spanner/BatchCreateSessionsTest.java (`testSpannerReturnsAllAvailableSessionsAndThenNoSessions`)

File: google-cloud-spanner/src/test/java/com/google/cloud/spanner/ReadAsyncTest.java
Comment (arpan14) [reply google-cloud-spanner/src/test/java/com/google/cloud/spanner/ReadAsyncTest.java:188]: Removed.

File: google-cloud-spanner/src/test/java/com/google/cloud/spanner/RetryOnInvalidatedSessionTest.java
Comment (arpan14) [reply google-cloud-spanner/src/test/java/com/google/cloud/spanner/RetryOnInvalidatedSessionTest.java:233]: Sure, sounds good.

File: google-cloud-spanner/src/test/java/com/google/cloud/spanner/SessionPoolTest.java
Comment (arpan14) [reply google-cloud-spanner/src/test/java/com/google/cloud/spanner/SessionPoolTest.java:2189]: Done

File: google-cloud-spanner/src/test/java/com/google/cloud/spanner/SpanTest.java
Comment (arpan14) [reply google-cloud-spanner/src/test/java/com/google/cloud/spanner/SpanTest.java:204]: I think both `SpanTest` and `OpenTelemetrySpanTest` exude a flaky behavior today irrespective of multiplexed sessions. What I saw is increasing this time reduces the flakiness.

File: google-cloud-spanner/src/test/java/com/google/cloud/spanner/SpannerGaxRetryTest.java
Comment (arpan14) [reply google-cloud-spanner/src/test/java/com/google/cloud/spanner/SpannerGaxRetryTest.java:218]: The regular sessions path uses `setMinSessions=0` as a workaround to prevent `BatchCreateSessions` RPC during `@Setup`. This ensures that during the test, the `BatchCreateSessions` RPC are invoked and they timeout.

For multiplexed session, the creation only happens during app initialization. Hence during `@Setup` itself the multiplexed session creation call will go. So there is no point of setting `setCreateSessionExecutionTime` within the test as by then mock spanner already would have received the call to create session and returned a response.

So given multiplexed session creation needs to be guaranteed within these tests, we would have to set the delay on whatever the next RPC is and then verify the GAX retry on the same.

File: google-cloud-spanner/src/test/java/com/google/cloud/spanner/SpannerGaxRetryTest.java
Comment (arpan14) [reply google-cloud-spanner/src/test/java/com/google/cloud/spanner/SpannerGaxRetryTest.java:138]: Makes sense. I have now modified the code to 
```
if (sessionPoolOptions.getUseMultiplexedSession()) {
      sessionPoolOptions.toBuilder().setWaitForMinSessions(Duration.ofSeconds(5));
    }
```

The previous code actually had no impact on regular sessions test because minSessions = 0. And hence the blocking wait time did not actually block for the regular sessions.

File: google-cloud-spanner/src/test/java/com/google/cloud/spanner/DatabaseClientImplTest.java
Comment (arpan14) [reply google-cloud-spanner/src/test/java/com/google/cloud/spanner/DatabaseClientImplTest.java:3016]: Removed.

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPool.java
Comment (arpan14) [reply google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPool.java:1789]: Yes, `getDatabaseClient(...).singleUse().read()` is completely non-blocking. So post app initialization multiple threads can get this object and issue a `next()` on the result set. When they do this there can be a race condition. Atleast some threads will get a `SettableApiFuture` which is not yet finished. And all of them will get blocked. 

In parallel another thread will issue the `CreateSession` call and will take a lock on `currentMultiplexedSessionReference` object in the `onSessesionReady()` callback function to initialize the session. This will unblock all the waiting threads. But I see the issue that in case the above thread are blocked on the uninitialized `SettableApiFuture` then that is already stale and these threads will eventually timeout.

However introducing a lock over here can introduce a deadlock? because some threads will acquire the lock on `currentMultiplexedSessionReference` and then `onSessesionReady` method will wait for this lock to be released. 

Do we need an additional waiter queue to mitigate this edge case?

Comment (olavloite) [reply google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPool.java]: I modified the implementation a bit to use a double-checked locking mechanism to ensure that only one thread will initialize the object. That way, we effectively create a 'waiter' queue for the multiplexed session instance.

File: google-cloud-spanner/src/test/java/com/google/cloud/spanner/ChannelUsageTest.java
Comment (olavloite) [reply google-cloud-spanner/src/test/java/com/google/cloud/spanner/ChannelUsageTest.java:226]: Are we using a random channel hint? If so, then I understand why it is flaky. The test iterates 2*numChannels and then checks that all channels have been used. That will often, but not always, be the case if we use random channel affiliation. The test will probably become acceptably non-flaky if we increase the number of iterations to something like 50*numChannels.

**Pull request merged**
